tornadox/solvers/ek1.py

```diff
@@ -0,0 +1,192 @@
+        state0,
+        *,
+        f,
+        df,
+        t1,
+        rtol,
+        atol,
+        safety=0.95,
+        factor_min=0.2,
+        factor_max=10.0,
+        power_integral_unscaled=0.3,
+        power_proportional_unscaled=0.4,
+    ):
+        """Perform a successful step."""
+
+        m0, c_sqrtm0 = state0.u
+        error_norm_previously_accepted = state0.error_norm
+
+        @jax.jit
+        def cond_fun(s):
+            _, state = s
+            return state.error_norm > 1
+
+        @jax.jit
+        def body_fun(s):
+            t, s_prev = s
+
+            # Never exceed the terminal value
+            dt_clipped = jnp.minimum(s_prev.dt_proposed, t1 - t0)
+            t_new = t0 + dt_clipped
+
+            # Compute preconditioner
+            p, p_inv = ibm.preconditioner_diagonal(
+                dt=dt_clipped, num_derivatives=num_derivatives
+            )
+            p = jnp.tile(p, ode_dimension)
+            p_inv = jnp.tile(p_inv, ode_dimension)
+
+            # Attempt step
+            x = ek1_projmatfree_d_nu_first_order.attempt_step_forward_only(
+                f=f,
+                df=df,
+                m=m0,
+                c_sqrtm=c_sqrtm0,
+                p=p,
+                p_inv=p_inv,
+                a=a,
+                q_sqrtm=q_sqrtm,
+                num_derivatives=num_derivatives,
+            )
+            (u_proposed, _, error) = x
+            error = dt_clipped * error  # The EK1 step does not do this. todo: why not?
+
+            # Normalise the error
+            m_proposed, _ = u_proposed
+            u1_ref = jnp.maximum(
+                jnp.abs(m_proposed[0 :: (num_derivatives + 1)]),
+                jnp.abs(m0[0 :: (num_derivatives + 1)]),
+            )
+            error_rel = error / (atol + rtol * u1_ref)
+            error_norm = jnp.linalg.norm(error_rel) / jnp.sqrt(error.size)
+
+            # Propose a new time-step
+            scale_factor = stepsizes.scale_factor_pi_control(
+                error_norm=error_norm,
+                error_order=num_derivatives + 1,
+                safety=safety,
+                factor_min=factor_min,
+                factor_max=factor_max,
+                error_norm_previously_accepted=error_norm_previously_accepted,
+                power_integral_unscaled=power_integral_unscaled,
+                power_proportional_unscaled=power_proportional_unscaled,
+            )
+            dt_new = scale_factor * dt_clipped
+            stats = s_prev.stats
+            stats["f_evaluation_count"] += 1
+            stats["df_evaluation_count"] += 1
+            stats["steps_attempted_count"] += 1
+
+            s_new = EK1State(
+                u=u_proposed, dt_proposed=dt_new, error_norm=error_norm, stats=stats
+            )
+            return t_new, s_new
+
+        larger_than_1 = 1.1
+        init_val = EK1State(
+            u=state0.u,
+            dt_proposed=state0.dt_proposed,
+            error_norm=larger_than_1,
```

Comment:
What is this for?

Reply:
the init_val? jax.lax.while_loop needs consistent types, shapes, etc. for the loop state.
the larger_than_1? Because if this is less than 1, the loop never actually "enters the loop", if that makes sense.

Reply:
I think I'm just a bit confused with all the body func while loop stuff, but I got it now. Thanks!

Reply:
Yes this is a little awkward, but cannot be avoided really. Let's keep this in mind and iterate whenever possible :)